### PR TITLE
Impl nc1

### PR DESCRIPTION
**Pull request by ColePBryan**
## Description

Adds initial implementation of the NCS_1 protocol from https://crypto.stanford.edu/~dabo/papers/netcode.pdf

closes: N/A

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests. !!! Needs more tests.....
- [x] Updated relevant documentation in the code
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -45,6 +45,7 @@ ark-ed-on-bls12-381 = { version = "^0.3.0", default-features = false, features =
 ark-bls12-377 = { version = "^0.3.0", default-features = false, features = [ "curve", "r1cs" ] }
 ark-mnt4-298 = { version = "^0.3.0", default-features = false, features = [ "curve", "r1cs" ] }
 ark-mnt6-298 = { version = "^0.3.0", default-features = false, features = [ "r1cs" ] }
+ark-bn254 = {git = "https://github.com/arkworks-rs/curves.git", rev = "99831650f8021cb6a16481bac674420bc6c1a5a1"}
```
Comment (tkornuta-semiotic) [inline Cargo.toml:48]: Alright! 👍

File: src/homomorphic_signature/mod.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        rng: &mut R,
+    ) -> Result<(Self::PublicKey, Self::SecretKey), Error>;
+
+    fn sign(
+        pp: &Self::Parameters,
+        sk: &Self::SecretKey,
+        tag: &[u8],
+        index: &[u8],
+        message: &[Self::Message],
+        //rng: &mut R, TODO: Research if this makes sense to have per message randomness in HSS signing
```
Comment (tkornuta-semiotic) [inline src/homomorphic_signature/mod.rs:31]: should we discuss this with Seve and Alexis?

Comment (aasseman) [reply src/homomorphic_signature/mod.rs:31]: Only Seve I think, that's outside of my expertise.

File: .gitignore [hunk 3]
```diff
@@ -9,4 +9,5 @@ Cargo.lock
 params
 *.swp
 *.swo
+.vscode
```
Comment (aasseman) [inline .gitignore]:
```suggestion
```

I wouldn't add ignores specific to your dev setup.

If you need that, do it with a global gitignore on your machine: https://stackoverflow.com/a/7335487

**Review by aasseman — CHANGES_REQUESTED** [review]
We'll also need a "deterministic" keygen that can take a 64 bytes seed (from a BIP-39 mnemonic).

File: src/homomorphic_signature/mod.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    fn single_sign_and_verify<S: HomomorphicSignatureScheme>(tag: &[u8], index: &[u8], message: &[S::Message]) {
+        let rng = &mut test_rng();
+        let n = message.len();
+        let parameters = S::setup(rng, n).unwrap();
+        let (pk, sk) = S::keygen(&parameters, rng).unwrap();
+        let sig = S::sign(&parameters, &sk, tag, index, message).unwrap();
+
+        assert!(S::verify(&parameters, &pk, tag, index, message, &sig).unwrap());
+    }
```
Comment (tkornuta-semiotic) [inline src/homomorphic_signature/mod.rs:78]: 1. not sure if I understand the split into the test and helper function in this case - especially that assert is done inside the helper
2. generally I prefer to separate tests from the code - but in here it is ok

**Review by tkornuta-semiotic — APPROVED** [review]
LGTM! Of course we could add more (unit) tests, but the integration test that covers both signing and verifying is sufficient for now.

File: src/homomorphic_signature/mod.rs (hunk 2 above)
Comment (ColePBryan) [reply src/homomorphic_signature/mod.rs:31]: Seve is aware of this comment. It is actually removed in the upcoming PR.

File: src/homomorphic_signature/mod.rs (hunk 4 above)
Comment (ColePBryan) [reply src/homomorphic_signature/mod.rs:78]: This test is written to match the style of the repo. There will also be other tests added. It seems to be convention in rust to put unit tests in the parent module file. Then Integration test will be in separate files. https://stackoverflow.com/questions/25107526/how-should-rust-unit-tests-be-organised

Comment (ColePBryan) [reply src/homomorphic_signature/mod.rs:78]: @tkornuta-semiotic let me know if we can resolve this as is, or if we need additional changes.

**Comment by aasseman** [comment]
2 remaining things on my side:

- > We'll also need a "deterministic" keygen that can take a 64 bytes seed (from a BIP-39 mnemonic).
  
  Can we, or do we already have that capability?
- Do an interactive rebase to cleanup the commit history (such as the .gitignore stuff) before merge.

**Review by aasseman — CHANGES_REQUESTED** [review]
2 remaining things on my side:

- > We'll also need a "deterministic" keygen that can take a 64 bytes seed (from a BIP-39 mnemonic).
  
  Can we, or do we already have that capability?
- Do an interactive rebase to cleanup the commit history (such as the .gitignore stuff) before merge.

File: src/homomorphic_signature/mod.rs (hunk 4 above)
Comment (tkornuta-semiotic) [reply src/homomorphic_signature/mod.rs:78]: this is a minor - I have accepted the PR already ;) Resolving.

**Comment by ColePBryan** [comment]
> 

We will likely implement an RNG to handle the deterministic keygen, that would keep the trait similar to the SignatureScheme trait. I added #2 to track the issue, we can implement that on another PR.

It is generally bad to do a rebase on code after pushed to remote. I can do so and force push it assuming no one else is using this code, but I would prefer not to start that precedent.

**Comment by aasseman** [comment]
> It is generally bad to do a rebase on code after pushed to remote. I can do so and force push it assuming no one else is using this code, but I would prefer not to start that precedent.

No, it's actually quite the opposite. The rule you are referring to applies only to main branches.

When the review process is done and we're ready to merge, a rebase is usually required to get up to the target's head, cleaning up commit history, and cleaning up commit messages.

https://akrabat.com/the-beginners-guide-to-rebasing-your-pr/
https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by ColePBryan** [comment]
@aasseman commits cleaned up, can you take a look and finish your review when you get a chance?

**Review by aasseman — APPROVED** [review]
LGTM!

**Pull request merged**